docs/arcadia-users-group/20240206-intro-to-formatting-and-linting/lesson.md

```diff
@@ -0,0 +1,433 @@
+
+```python
+# create a random timelapse image with 3 timepoints
+image = np.random.rand(3, 100, 100)
+
+# iterate over the pixels in the image
+N_T, N_X, N_Y = image.shape
+for t in range(N_T):
+    for x in range(N_X):
+        for y in range(N_Y):
+            v = image[t, x, y]
+            print(f'The pixel intensity at time {t} and position ({x}, {y}) is {v}')
+```
+Although this code is compact, it is hard to keep track of the many single-character variables, and it will only become harder as the body of the nested for loops grows more complex. In addition, although they are not literally single characters, the names `N_X` and `N_Y` are ambiguous (and also wrongly capitalized). 
+
+This version of the code, with more descriptive variable names, is much clearer:
+```python
+# create a random timelapse image with 3 timepoints
+image = np.random.rand(3, 100, 100)
+
+# iterate over the pixels in the image
+num_timepoints, num_rows, num_cols = image.shape
+for time_ind in range(num_timepoints):
+    for row_ind in range(num_rows):
+        for col_ind in range(num_cols):
+            intensity = image[time_ind, row_ind, col_ind]
+            print(
+                f'The pixel intensity at timepoint {time_ind} '
+                f'and position ({row_ind}, {col_ind}) is {intensity}'
+            )
+```
+</details>
+
+### Documentation conventions
+In the context of code style guides, "documentation" refers to human-readable text that is embedded in the source code to explain what the code does and how it works. It is very important to define standards and conventions for documentation because it is a major way--and sometimes the only way--to ensure that code is readily understandable by others. In Python, documentation takes two forms: comments that can appear anywhere in the code and docstrings that accompany modules, classes, and functions.
+
+#### Comments
+Comments are human-readable lines of text that can appear anywhere in the source code. In Python, they are denoted by the `#` character. Although they are ignored by the Python interpreter and might seem like an area where "anything goes", it is important to adhere to strict standards of grammar and style when writing comments, since they are often the only way to explain the purpose of a particular line or block of code.
+
+In particular, comments should be complete sentences with proper punctuation. They should be written in full English sentences and should be grammatically correct. Importantly, they should end with periods; this is not only grammatically correct but is also the only way to indicate to the reader that the comment is complete and was not accidentally truncated (or never completely written in the first place). 
+
+Most importantly, comments should generally be used to explain *why* a particular line or block of code is doing what it is doing, not *what* it is doing or *how* it is doing it (usually, this should be apparent from the code itself). 
```

Comment:
I'd probably expand this a little bit and say comments are good for drawing attention, so while the "why" is generally important, in cases where you're doing something unintuitive/weird/magical, it could be worth calling out the what as well as the why.